services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java

```diff
@@ -0,0 +1,345 @@
+      for (FieldDescriptor extension : type.getExtensions()) {
+        processExtension(extension, fd);
+      }
+      for (Descriptor nestedType : type.getNestedTypes()) {
+        processType(nestedType, fd);
+      }
+    }
+
+    private void processFileDescriptor(FileDescriptor fd) {
+      String fdName = fd.getName();
+      Preconditions.checkState(!fileDescriptorsByName.containsKey(fdName),
+          "File name already used: %s", fdName);
+      fileDescriptorsByName.put(fdName, fd);
+      for (ServiceDescriptor service : fd.getServices()) {
+        processService(service, fd);
+      }
+      for (Descriptor type : fd.getMessageTypes()) {
+        processType(type, fd);
+      }
+      for (FieldDescriptor extension : fd.getExtensions()) {
+        processExtension(extension, fd);
+      }
+    }
+
+    /**
+     * Processes the service file descriptors. This prepares a list of the service names and
+     * indexes the file descriptors by name, symbol, and extension.
+     */
+    private void initFileDescriptorMaps() {
+      serviceNames = new HashSet<String>();
+      fileDescriptorsByName = new HashMap<String, FileDescriptor>();
+      fileDescriptorsBySymbol = new HashMap<String, FileDescriptor>();
+      fileDescriptorsByExtensionAndNumber = new HashMap<String, Map<Integer, FileDescriptor>>();
+
+      List<ServerServiceDefinition> services = server.getServices();
+
+      Queue<FileDescriptor> fileDescriptorsToProcess = new LinkedList<FileDescriptor>();
+      Set<String> seenFiles = new HashSet<String>();
+      for (ServerServiceDefinition service : services) {
+        io.grpc.ServiceDescriptor serviceDescriptor = service.getServiceDescriptor();
+        if (serviceDescriptor.getMarshallerDescriptor() instanceof ProtoFileDescriptorSupplier) {
+          FileDescriptor fileDescriptor =
+              ((ProtoFileDescriptorSupplier) serviceDescriptor.getMarshallerDescriptor())
+                  .getFileDescriptor();
+          String serviceName = serviceDescriptor.getName();
+          Preconditions.checkState(!serviceNames.contains(serviceName),
+              "Service already defined: %s", serviceName);
+          serviceNames.add(serviceName);
+          if (!seenFiles.contains(fileDescriptor.getName())) {
+            seenFiles.add(fileDescriptor.getName());
+            fileDescriptorsToProcess.add(fileDescriptor);
+          }
+        }
+      }
+
+      while (!fileDescriptorsToProcess.isEmpty()) {
+        FileDescriptor currentFd = fileDescriptorsToProcess.remove();
+        processFileDescriptor(currentFd);
+        for (FileDescriptor dependencyFd : currentFd.getDependencies()) {
+          if (!seenFiles.contains(dependencyFd.getName())) {
+            seenFiles.add(dependencyFd.getName());
+            fileDescriptorsToProcess.add(dependencyFd);
+          }
+        }
+      }
+    }
+
+    @Override
+    public void onNext(ServerReflectionRequest request) {
+      initFileDescriptorMaps();
```

Comment:
I still don't get why this isn't loaded in the constructor. We can deal with it later in a followup PR.